fpga_interchange/examples/chipdb.cmake

```diff
@@ -0,0 +1,314 @@
+
+    if(DEFINED device_target)
+        set(${device_target} ${constraints_luts_device} PARENT_SCOPE)
+    endif()
+endfunction()
+
+function(generate_chipdb)
+    # ~~~
+    # generate_chipdb(
+    #    family <family>
+    #    device <common device>
+    #    part <part>
+    #    device_target <device target>
+    #    bel_bucket_seeds <bel bucket seeds>
+    #    package <package>
+    # )
+    # ~~~
+    #
+    # Generates a chipdb BBA file, starting from a device database.
+    #
+    # The chipdb binary file is directly generated to the
+    # <nextpnr-root>/build/fpga_interchange/chipdb/ directory.
+    #
+    # The package argument is only used to run the architecture check target.
+    #
+    # Targets generated:
+    #   - chipdb-${device}-bba
+    #   - chipdb-${device}-bin
+    #   - device-${device}
+    #
+    # The device-${device} target contains properties to get the interchange device as well
+    # as the binary chipdb
+
+    set(options)
+    set(oneValueArgs family device part device_target bel_bucket_seeds package)
+    set(multiValueArgs)
+
+    cmake_parse_arguments(
+        generate_chipdb
+        "${options}"
+        "${oneValueArgs}"
+        "${multiValueArgs}"
+        ${ARGN}
+    )
+
+    set(family ${generate_chipdb_family})
+    set(device ${generate_chipdb_device})
+    set(part ${generate_chipdb_part})
+    set(device_target ${generate_chipdb_device_target})
+    set(bel_bucket_seeds ${generate_chipdb_bel_bucket_seeds})
+    set(package ${generate_chipdb_package})
+
+    get_target_property(device_loc ${device_target} LOCATION)
+    set(chipdb_bba ${CMAKE_CURRENT_BINARY_DIR}/chipdb.bba)
+    add_custom_command(
+        OUTPUT ${chipdb_bba}
+        COMMAND
+            ${PYTHON_EXECUTABLE} -mfpga_interchange.nextpnr_emit
+                --schema_dir ${INTERCHANGE_SCHEMA_PATH}
+                --output_dir ${CMAKE_CURRENT_BINARY_DIR}
+                --bel_bucket_seeds ${bel_bucket_seeds}
+                --device ${device_loc}
+        DEPENDS
+            ${bel_bucket_seeds}
+            ${device_target}
+            ${device_loc}
+    )
+
+    add_custom_target(chipdb-${device}-bba DEPENDS ${chipdb_bba})
+
+    set(chipdb_bin ${chipdb_dir}/chipdb-${device}.bin)
+    add_custom_command(
+        OUTPUT ${chipdb_bin}
+        COMMAND
+            bbasm -l ${chipdb_bba} ${chipdb_bin}
+        DEPENDS
+            chipdb-${device}-bba
+            ${chipdb_bba}
+            bbasm
+    )
+
+    add_custom_target(chipdb-${device}-bin DEPENDS ${chipdb_bin})
+
+    # Setting device target properties
+    add_custom_target(device-${device})
+    set_target_properties(
+        device-${device}
+        PROPERTIES
+            DEVICE_LOC ${device_loc}
+            DEVICE_TARGET ${device_target}
+            CHIPDB_BIN_LOC ${chipdb_bin}
+            CHIPDB_BIN_TARGET chipdb-${device}-bin
+    )
+
+    # Generate architecture check target
+    set(test_data_source ${CMAKE_CURRENT_SOURCE_DIR}/test_data.yaml)
```

Comment:
Why is this symlink needed?  I believe this will prevent in-source builds from working properly.

Reply:
This is required as otherwise we get:

```
Traceback (most recent call last):
  File "/data/interchange/nextpnr/python/check_arch_api.py", line 90, in <module>
    check_arch_api(ctx)
  File "/data/interchange/nextpnr/python/check_arch_api.py", line 64, in check_arch_api
    with open('test_data.yaml', 'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'test_data.yaml'
ERROR: Error occurred while executing Python script /data/interchange/nextpnr/python/check_arch_api.py
0 warnings, 1 error
fpga_interchange/examples/devices/xc7a35t/CMakeFiles/chipdb-xc7a35t-bin-check-test-data.dir/build.make:57: recipe for target 'fpga_interchange/examples/devices/xc7a35t/CMakeFiles/chipdb-xc7a35t-bin-check-test-data' failed
```

test_data.yaml needs to reside in the build directory, hence the symlink

Reply:
Just have the working directory of the command in the source directory where the tile is?

Reply:
Right, done